README.md
# CloudFormation Launcher (CFNL) BASH Script

## Purpose
Used to launch CloudFormation Template from command line using AWS CLI commands.

It is designed to be reusable without having to make changes to the script. 
Simple use a default create YAML config or create you own custom YAML config file.

## Tested Environment
* macOS Sierra 10.12.3
* AWS CLI v1.11.55

## Prerequisites
This has been test
* BASH Shell
* **tee** command installed
    * Unless set ```nolog: true``` in config
* Proxy configured for Shell if needed
* AWS CLI installed and configured

## Setup
1. Open Terminal
2. Pull down this git repository locally
    1. ```git clone https://github.com/bonusbits/cfn_launcher.git /path/to/clone/```
3. Create symlink & aliases the the *cfnl.sh* as desired. Examples below.

    
## Help
The following can be displayed with the *-h* switch.
    
```
CloudFormation Launcher v1.7.0

Usage: /usr/local/bin/cfnl [-u | -d | -s] -f ./config_file.yml

Options:
    -f File Path             :  (Required) YAML Script Config File Full Path
    -r Region                :  Overrides yaml config, ENV:AWS_REGION and aws config.
                                Set Order: 1. Parameter 2. Yaml Config 3. ENV:AWS_REGION 4. Default us-west-2
    -u Action Update         :  Action Flag to Update Stack
    -d Action Delete         :  Action Flag to Delete Stack
    -s Action Status         :  Action Flag to Get Stack Status
    -b Debug Output          :  Display Additional Output for Debugging
    -h Help                  :  Displays Help Information
    -v Version               :  Displays Script Version

Action Flags:
    Only one action flag can be used. The default Action is 'Create' (No flag).
    The three override Action Flags are -u, -d and -s.

Description:
    This script uses the AWS CLI and BASH to create, update, delete or get
    status of a CloudFormation Stack. It uses the AWS CLI to push the
    CloudFormation Template to AWS. Then loops over and over checking
    status of the stack.

YAML Config File Format Example:
    stackname: stack1
    profilename: awsaccount
    region: us-west-2
    templateurl: https://s3.amazonaws.com/bucket/webapp1.yml # Or .json
    templatelocal: /path/to/cfn/templates/webapp1.yml # Unless using URL
    parametersfilepath: /Users/gsm987/.cfnl/uswest2/client1/account1/dev/webapp1.json
    capabilityiam: false
    capabilitynamediam: false
    deletecreatefailures: true
    uses3template: true
    nolog: false
    logfile: /Users/gsm987/.cfnl/logs/uswest2/client1/account1/dev/webapp1.log
    verbose: true
    waittime: 5
    maxwaits: 180

Examples:
    Create Stack
    /usr/local/bin/cfnl -f /Users/gsm987/.cfnl/uswest2/client1/account1/dev/webapp1.yml

    Update Stack
    /usr/local/bin/cfnl -u -f /Users/gsm987/.cfnl/uswest2/client1/account1/dev/webapp1.yml

    Delete Stack
    /usr/local/bin/cfnl -d -f /Users/gsm987/.cfnl/uswest2/client1/account1/dev/webapp1.yml

    Stack Status
    /usr/local/bin/cfnl -s -f /Users/gsm987/.cfnl/uswest2/client1/account1/dev/webapp1.yml

    Create Stack with Debugging
    /usr/local/bin/cfnl -b -f /Users/gsm987/.cfnl/uswest2/client1/account1/dev/webapp1.yml

    Delete Stack with Debugging
    /usr/local/bin/cfnl -d -b -f /Users/gsm987/.cfnl/uswest2/client1/account1/dev/webapp1.yml

Author:
    Levon Becker
    https://github.com/LevonBecker
    https://www.bonusbits.com
```    
    
## Example BASH Profile Functions and Aliases
Examples BASH Profile scripts to simplify use, but are not required of course.

### .bash_profile
```bash
# CloudFormation Launcher Aliases
if [ -f $HOME/.bash_cfnl ]; then
	  source $HOME/.bash_cfnl
fi
```

### .bash_cfnl
```bash
#!/usr/bin/env bash

# CloudFormation Launcher - https://github.com/bonusbits/cfn_launcher

# Symlink
if [ ! -h "/usr/local/bin/cfnl" ]; then
   ln -s "$HOME/Development/github/bonusbits/cfn_launcher/cfnl.sh" /usr/local/bin/cfnl
fi

function cfnl-show() {
    echo "CFNL_PATH: ($CFNL_PATH)"
}

function cfnl-set-path() {
    cfnl_configs_path=$1

    export CFNL_PATH="$cfnl_configs_path"
    cfnl-show
}

# Default
cfnl-set-path $HOME/.cfnl/uswest2/client1/account01/prd/project01

function cfc() {
    # Create
    /usr/local/bin/cfnl -f "$CFNL_PATH/$1.yml"
}

function cfu() {
    # Update
    /usr/local/bin/cfnl -u -f "$CFNL_PATH/$1.yml"
}

function cfd() {
    # Delete
    /usr/local/bin/cfnl -d -f "$CFNL_PATH/$1.yml"
}

function cfs() {
    # Status
    /usr/local/bin/cfnl -s -f "$CFNL_PATH/$1.yml"
}

function cfcd() {
    # Create with Debug
    /usr/local/bin/cfnl -b -f "$CFNL_PATH/$1.yml"
}

function cfud() {
    # Update with Debug
    /usr/local/bin/cfnl -b -u -f "$CFNL_PATH/$1.yml"
}

function cfdd() {
    # Delete with Debug
    /usr/local/bin/cfnl -b -d -f "$CFNL_PATH/$1.yml"
}

function cfsd() {
    # Status with Debug
    /usr/local/bin/cfnl -b -s -f "$CFNL_PATH/$1.yml"
}
```

### Example Folder Structure
You can use whatever folder structure you like. This is just an example.

```
 $HOME/.cfnl/
└── uswest2
    └── client01
        └── account01
            ├── dev
            │   ├── project01
            │   │   └── efs.yml
            │   │   └── rds-snapshot.yml
            │   │   └── web.yml
            │   ├── shared
            │   │   └── vpc.yml
            │   │   └── utm.yml
            │   └── project02
            │       ├── efs.yml
            │       └── rds-snapshot.yml
            │       └── web.yml
            └── prd
                ├── project01
                │   ├── efs.yml
                │   ├── rds.yml
                │   └── web.yml
                ├── shared
                │   ├── vpc.yml
                │   └── utm.yml
                └── project02
                    ├── efs.yml
                    ├── rds.yml
                    ├── web-blue.yml
                    └── web-green.yml
```

### Example Aliased Commands
```bash
# Set CFNL Path
cfnl-set-path $HOME/.cfnl/uswest2/client01/account01/dev/project01
# CFNL is now set to use $HOME/.cfnl/uswest2/client01/account01/dev/project01/*.yml

# Create Stack
cfc web
# Delete Stack
cfd web
# Update Stack
cfu web
# Stack Status
cfs web
```
    
## Output Examples
### Create
    [2017-03-21_13:44:35] ** Start CloudFormation Launcher v1.6.0 **
    [2017-03-21_13:44:35] ACTION: CREATE
    {
        "StackId": "arn:aws:cloudformation:us-west-2:000000000000:stack/webapp1/32046180-0e78-11e7-86c8-513ac9841b19"
    }
    [2017-03-21_13:44:36] REPORT: Successfully Executed CREATE Stack Command
    
    [2017-03-21_13:44:37] REPORT: Successfully Executing Status Check
    [2017-03-21_13:44:37] REPORT: Status (CREATE_IN_PROGRESS)
    [2017-03-21_13:44:37] REPORT: CREATE stack is not complete!
    [2017-03-21_13:44:37] REPORT: Attempt 1 of 360.
    [2017-03-21_13:44:37] RUNTIME: 00 minutes 02 seconds
    [2017-03-21_13:44:37] REPORT: Polling again in 5 seconds...
    
    [2017-03-21_13:44:42] REPORT: Successfully Executing Status Check
    [2017-03-21_13:44:42] REPORT: Status (CREATE_IN_PROGRESS)
    [2017-03-21_13:44:42] REPORT: CREATE stack is not complete!
    [2017-03-21_13:44:42] REPORT: Attempt 2 of 360.
    [2017-03-21_13:44:42] RUNTIME: 00 minutes 07 seconds
    [2017-03-21_13:44:42] REPORT: Polling again in 5 seconds...
    
    [2017-03-21_13:44:48] REPORT: Successfully Executing Status Check
    [2017-03-21_13:44:48] REPORT: Status (CREATE_IN_PROGRESS)
    [2017-03-21_13:44:48] REPORT: CREATE stack is not complete!
    [2017-03-21_13:44:48] REPORT: Attempt 3 of 360.
    [2017-03-21_13:44:48] RUNTIME: 00 minutes 13 seconds
    [2017-03-21_13:44:48] REPORT: Polling again in 5 seconds...
    
    ... Fast Forward ...
    
    [2017-03-21_14:00:48] REPORT: Successfully Executing Status Check
    [2017-03-21_14:00:48] REPORT: Status (CREATE_IN_PROGRESS)
    [2017-03-21_14:00:48] REPORT: CREATE stack is not complete!
    [2017-03-21_14:00:48] REPORT: Attempt 172 of 360.
    [2017-03-21_14:00:48] RUNTIME: 16 minutes 13 seconds
    [2017-03-21_14:00:48] REPORT: Polling again in 5 seconds...
    
    [2017-03-21_14:00:54] REPORT: Successfully Executing Status Check
    [2017-03-21_14:00:54] REPORT: Status (CREATE_COMPLETE)
    [2017-03-21_14:00:54] REPORT: CREATE Completed!
    
    [2017-03-21_14:00:54] ENDTIME: (Tue Mar 21 14:00:54 PDT 2017)
    [2017-03-21_14:00:54] RUNTIME: 16 minutes 19 seconds
    
    [2017-03-21_14:00:54] REPORT: CREATE SUCCESS!

### Update
    [2017-03-21_14:11:10] ** Start CloudFormation Launcher v1.6.0 **
    [2017-03-21_14:11:10] ACTION: UPDATE
    {
        "StackId": "arn:aws:cloudformation:us-west-2:000000000000:stack/webapp1/32046180-0e78-11e7-86c8-513ac9841b19"
    }
    [2017-03-21_14:11:11] REPORT: Successfully Executed UPDATE Stack Command
    
    [2017-03-21_14:11:12] REPORT: Successfully Executing Status Check
    [2017-03-21_14:11:12] REPORT: Status (UPDATE_IN_PROGRESS)
    [2017-03-21_14:11:12] REPORT: UPDATE stack is not complete!
    [2017-03-21_14:11:12] REPORT: Attempt 1 of 360.
    [2017-03-21_14:11:12] RUNTIME: 00 minutes 02 seconds
    [2017-03-21_14:11:12] REPORT: Polling again in 5 seconds...
    
    [2017-03-21_14:11:17] REPORT: Successfully Executing Status Check
    [2017-03-21_14:11:17] REPORT: Status (UPDATE_IN_PROGRESS)
    [2017-03-21_14:11:17] REPORT: UPDATE stack is not complete!
    [2017-03-21_14:11:17] REPORT: Attempt 2 of 360.
    [2017-03-21_14:11:17] RUNTIME: 00 minutes 07 seconds
    [2017-03-21_14:11:17] REPORT: Polling again in 5 seconds...
    
    ... Fast Forward ...
    
    [2017-03-21_14:11:46] REPORT: Successfully Executing Status Check
    [2017-03-21_14:11:46] REPORT: Status (UPDATE_IN_PROGRESS)
    [2017-03-21_14:11:46] REPORT: UPDATE stack is not complete!
    [2017-03-21_14:11:46] REPORT: Attempt 7 of 360.
    [2017-03-21_14:11:46] RUNTIME: 00 minutes 36 seconds
    [2017-03-21_14:11:46] REPORT: Polling again in 5 seconds...
    
    [2017-03-21_14:11:52] REPORT: Successfully Executing Status Check
    [2017-03-21_14:11:52] REPORT: Status (UPDATE_COMPLETE)
    [2017-03-21_14:11:52] REPORT: UPDATE Completed!
    
    [2017-03-21_14:11:52] ENDTIME: (Tue Mar 21 14:11:52 PDT 2017)
    [2017-03-21_14:11:52] RUNTIME: 00 minutes 42 seconds
    
    [2017-03-21_14:11:52] REPORT: UPDATE SUCCESS!

### Status
    /Users/username/.cfnl/uswest2/client1/account01/prd/webapp1.yml
    Stack Status: (UPDATE_COMPLETE)
    
### Delete
    [2017-03-21_14:17:20] ** Start CloudFormation Launcher v1.6.0 **
    [2017-03-21_14:17:20] ACTION: DELETE
    [2017-03-21_14:17:20] ACTION: Deleting Stack
    [2017-03-21_14:17:21] REPORT: Successfully Executed Delete Stack Command
    [2017-03-21_14:17:21] REPORT: Successfully Executing Status Check
    [2017-03-21_14:17:21] REPORT: Status (DELETE_IN_PROGRESS)
    [2017-03-21_14:17:21] REPORT: Delete not complete!
    [2017-03-21_14:17:21] REPORT: Attempt 1 of 360.
    [2017-03-21_14:17:21] RUNTIME: 00 minutes 01 seconds
    [2017-03-21_14:17:21] Polling again in 5 seconds...
    
    [2017-03-21_14:17:27] REPORT: Successfully Executing Status Check
    [2017-03-21_14:17:27] REPORT: Status (DELETE_IN_PROGRESS)
    [2017-03-21_14:17:27] REPORT: Delete not complete!
    [2017-03-21_14:17:27] REPORT: Attempt 2 of 360.
    [2017-03-21_14:17:27] RUNTIME: 00 minutes 07 seconds
    [2017-03-21_14:17:27] Polling again in 5 seconds...
    
    ... Fast Forward ...
    
    [2017-03-21_14:18:54] REPORT: Successfully Executing Status Check
    [2017-03-21_14:18:54] REPORT: Status (DELETE_IN_PROGRESS)
    [2017-03-21_14:18:54] REPORT: Delete not complete!
    [2017-03-21_14:18:54] REPORT: Attempt 17 of 360.
    [2017-03-21_14:18:54] RUNTIME: 01 minutes 34 seconds
    [2017-03-21_14:18:54] Polling again in 5 seconds...
    
    [2017-03-21_14:19:00] REPORT: Successfully Executing Status Check
    [2017-03-21_14:19:00] REPORT: Status (DELETE_IN_PROGRESS)
    [2017-03-21_14:19:00] REPORT: Delete not complete!
    [2017-03-21_14:19:00] REPORT: Attempt 18 of 360.
    [2017-03-21_14:19:00] RUNTIME: 01 minutes 40 seconds
    [2017-03-21_14:19:00] Polling again in 5 seconds...
    
    
    An error occurred (ValidationError) when calling the DescribeStacks operation: Stack with id webapp1 does not exist
    [2017-03-21_14:19:06] REPORT: Successfully Executing Status Check
    [2017-03-21_14:19:06] REPORT: Status (DOES NOT EXIST)
    [2017-03-21_14:19:06] REPORT: Stack Deleted (webapp1)
    
    [2017-03-21_14:19:06] ENDTIME: (Tue Mar 21 14:19:06 PDT 2017)
    [2017-03-21_14:19:06] RUNTIME: 01 minutes 46 seconds
    
    [2017-03-21_14:19:06] REPORT: DELETE SUCCESS!

### Status
    /Users/username/.cfnl/uswest2/client1/account01/prd/webapp1.yml
    Stack Status: (DOES NOT EXIST)
    
## Troubleshooting
* Can not use underscores of hyphens in yaml properties file key names.


CHANGELOG.md
# CHANGE LOG

## 1.7.0 - 05/04/2017 - Levon Becker
* Added create log path if missing.
* Added aws region to be included in the aws cli calls
    * Added parameter to set region and override others
    * Added yaml config region: key/value check
    * Third option that is used for the region is to use ENV:AWS_REGION
    * Last is to default to a region us-west-2
    * When switching around between regions your .aws/config could be hard set to a region and is a headache to have to change the config when it's plenty easy to add as argument or just use ENV var so more dynamic.
* Renamed REPORT to INFO for logging
* Added some more debug messages
* Added status show_header section
* Moved start and add debug enable messages to be before any functions are run outside of show_header
* Remove Dev Null for Status check because apparently exit 255 can be that the stack doesn't exist, but also if your STS Token has expired.. lame that it's not a different exit code.

## 1.6.1 - 03/24/2017 - Levon Becker
* Added Config to regular header output

## 1.6.0 - 03/21/2017 - Levon Becker
* Added Stack Status Option
* Added Error Handling for multiple Action Flag Options
* Renamed Script from cfn-launcher.sh to cfnl.sh

## 1.5.1 - 11/15/2016 - Levon Becker
* Removed Task Type from header since Action replaces it.
* Move Display Runtime to Function so can be used during loops
* Improved Runtime output

## 1.5.0 - 11/15/2016 - Levon Becker
* Switched yaml parameter file argument from -c (config) to -f (file) So it doesn't get confused with "create"

## 1.4.0 - 11/15/2016 - Levon Becker
* Added Delete Stack switch argument and logic
* Added --profile to describe command
* Merged exit check functions
* Added delete triggered logic for exit check when the stack is removed a 255 will return
* Added no log option to yaml so log file write can be disabled
* Removed date & time from version
* Merged message functions
* Moved Action variable logic
* Added Action output to Header
* Added No Header Option

## 1.3.0 - 11/02/2016 - Levon Becker
* Renamed iamaccess to capabilityiam
* Added capabilitynamediam yaml config option

## 1.2.6 - 11/02/2016 - Stable - Levon Becker
* Added Version switch argument to display version

## 1.2.5 - 11/01/2016 - Stable - Levon Becker
* Added missing character in var for CAPABILITY_IAM

## 1.2.4 - 10/27/2016 - Stable - Levon Becker
* Added timeout count option for yaml config
* Fixed indents to 4 spaces
* Moved some of the loop logic to functions to reduce complexity

## 1.2.3 - 10/26/2016 - Stable - Levon Becker
* Fixed some conditions to be posix compliant

## 1.2.1 - 10/24/2016 - Stable - Levon Becker
* Added Help Message

## 1.2.0 - 10/24/2016 - Stable - Levon Becker
* Removed createstack property and made create stack default with -u switch to trigger update
* Fixed usage message

## 1.1.4 - 10/24/2016 - Stable - Levon Becker
* Fixed loop to work correctly with UPDATE_COMPLETE_CLEANUP_IN_PROGRESS status
* Fixed a couple strings that needed to be interpreted for console action status
* Changed script version

## 1.1.3 - 10/24/2016 - Levon Becker
* Changed argument switch to -c for config instead of -p for properties
* Updated Readme
* Created a create and update example config and moved to own directory

## 1.1.2 - 10/24/2016 - Levon Becker
* Fixed Header Logging to file
* Change binary name to hyphen instead of underscore

## 1.1.1 - 10/20/2016 - Levon Becker
* Added verbose option to get create complete outputs on fail so can track what worked and where it failed.

## 1.1.0 - 10/18/2016 - Stable - Levon Becker
* Added YAML Properties File Parser

## 1.1.2 - 10/17/2016 - Beta - Levon Becker
* WIP

## 1.0.1 - 10/14/2016 - Beta - Levon Becker
* WIP

## 1.0.0 - 10/13/2016 - Skeleton - Levon Becker
* Initial Commit

- - -

cfnl.sh
#!/usr/bin/env bash

# Static Variables
successful=false
delete_successful=false
triggered_delete=false
script_version=1.7.0
# unset stack_name
# read -p "Enter Stack Name: " stack_name

function help_message () {
helpmessage="Description:
    This script uses the AWS CLI and BASH to create, update, delete or get
    status of a CloudFormation Stack. It uses the AWS CLI to push the
    CloudFormation Template to AWS. Then loops over and over checking
    status of the stack.

YAML Config File Format Example:
    stackname: stack1
    profilename: awsaccount
    region: us-west-2
    templateurl: https://s3.amazonaws.com/bucket/webapp1.yml # Or .json
    templatelocal: /path/to/cfn/templates/webapp1.yml # Unless using URL
    parametersfilepath: $HOME/.cfnl/uswest2/client1/account1/dev/webapp1.json
    capabilityiam: false
    capabilitynamediam: false
    deletecreatefailures: true
    uses3template: true
    nolog: false
    logfile: $HOME/.cfnl/logs/uswest2/client1/account1/dev/webapp1.log
    verbose: true
    waittime: 5
    maxwaits: 180

Examples:
    Create Stack
    $0 -f $HOME/.cfnl/uswest2/client1/account1/dev/webapp1.yml

    Update Stack
    $0 -u -f $HOME/.cfnl/uswest2/client1/account1/dev/webapp1.yml

    Delete Stack
    $0 -d -f $HOME/.cfnl/uswest2/client1/account1/dev/webapp1.yml

    Stack Status
    $0 -s -f $HOME/.cfnl/uswest2/client1/account1/dev/webapp1.yml

    Create Stack with Debugging
    $0 -b -f $HOME/.cfnl/uswest2/client1/account1/dev/webapp1.yml

    Delete Stack with Debugging
    $0 -d -b -f $HOME/.cfnl/uswest2/client1/account1/dev/webapp1.yml

Author:
    Levon Becker
    https://github.com/LevonBecker
    https://www.bonusbits.com
"
    usage
    echo "$helpmessage";
}

function version_message() {
versionmessage="CloudFormation Launcher v$script_version"
    echo "$versionmessage";
}

function usage() {
usagemessage="Usage: $0 [-u | -d | -s] -f ./config_file.yml

Options:
    -f File Path             :  (Required) YAML Script Config File Full Path
    -r Region                :  Overrides yaml config, ENV:AWS_REGION and aws config.
                                Set Order: 1. Parameter 2. Yaml Config 3. ENV:AWS_REGION 4. Default us-west-2
    -u Action Update         :  Action Flag to Update Stack
    -d Action Delete         :  Action Flag to Delete Stack
    -s Action Status         :  Action Flag to Get Stack Status
    -b Debug Output          :  Display Additional Output for Debugging
    -h Help                  :  Displays Help Information
    -v Version               :  Displays Script Version

Action Flags:
    Only one action flag can be used. The default Action is 'Create' (No flag).
    The three override Action Flags are -u, -d and -s.
"
    version_message
    echo ''
    echo "$usagemessage";
}

while getopts "f:r:bdhsuv" opts; do
    case $opts in
        b ) debug=true;;
        d ) delete=true;;
        f ) config_file_path=$OPTARG;;
        r ) opt_region=$OPTARG;;
        h ) help_message; exit 0;;
        s ) status=true;;
        u ) update=true;;
        v ) version_message; exit 0;;
    esac
done

action_flag_count=0
if [ -n "$delete" ]; then
    action_flag_count=$[$action_flag_count +1]
    flags_used+="-d "
fi
if [ -n "$update" ]; then
    action_flag_count=$[$action_flag_count +1]
    flags_used+="-u "
fi
if [ -n "$status" ]; then
    action_flag_count=$[$action_flag_count +1]
    flags_used+="-s "
fi
if [ "$action_flag_count" -gt 1 ]; then
    usage
    echo "ERROR: Multiple Action Flags detected! ($flags_used)"
    exit 1
fi

if [ "$config_file_path" == "" ]; then
    usage
    echo 'ERROR: A YAML Config File is Required!'
    exit 1
fi

# Set Task Type
if [ "$update" == "true" ]; then
    task_type=update-stack
else
    task_type=create-stack
fi

function parse_yaml() {
   local prefix=$2
   local s='[[:space:]]*' w='[a-zA-Z0-9_]*' fs=$(echo @|tr @ '\034')
   sed -ne "s|^\($s\)\($w\)$s:$s\"\(.*\)\"$s\$|\1$fs\2$fs\3|p" \
        -e "s|^\($s\)\($w\)$s:$s\(.*\)$s\$|\1$fs\2$fs\3|p"  $1 |
   awk -F$fs '{
      indent = length($1)/2;
      vname[indent] = $2;
      for (i in vname) {if (i > indent) {delete vname[i]}}
      if (length($3) > 0) {
         vn=""; for (i=0; i<indent; i++) {vn=(vn)(vname[i])("_")}
         printf("%s%s%s=\"%s\"\n", "'$prefix'",vn, $2, $3);
      }
   }'
}

function message() {
    DATETIME=$(date +%Y-%m-%d_%H:%M:%S)
    if [ "$yaml_nolog" == "true" ]; then
        echo "[$DATETIME] $*"
    else
        echo "[$DATETIME] $*" | tee -a ${yaml_logfile}
    fi
}

function show_header {
    if [ "$yaml_uses3template" == "true" ]; then
        TEMPLATE=${yaml_templateurl}
    else
        TEMPLATE=${yaml_templatelocal}
    fi

    # If just getting Status things like ACTION are not set, so need own header
	if [ "$status" == "true" ]; then
        message "ACTION: STATUS"
        message "CONFIG: ($config_file_path)"
        message "STACKNAME: ($yaml_stackname)"
	elif [ "$debug" == "true" ]; then
        message '** PARAMETERS **'
        message "ACTION: $ACTION"
        message "STACK NAME: $yaml_stackname"
        message "PROFILE: $yaml_profilename"
        message "REGION: $region"
        message "TEMPLATE: $TEMPLATE"
        message "PARAMETERS FILE: $yaml_parametersfilepath"
        message "CAPABILITY IAM: $yaml_capabilityiam"
        message "CAPABILITY NAMED IAM: $yaml_capabilitynamediam"
        message "NO LOG: $yaml_nolog"
        message "LOG FILE: $yaml_logfile"
        message "VERBOSE: $yaml_verbose"
        message "LAUNCHER CONFIG: $config_file_path"
        message "DELETE ON FAILURE: $yaml_deletecreatefailures"
        message "WAIT TIME (Sec): $yaml_waittime"
        message "MAX WAITS (Loops): $yaml_maxwaits"
	else
        message "ACTION: $ACTION"
        message "CONFIG: $config_file_path"
	fi
}

function exit_check {
    if [ "$triggered_delete" == "true" ]; then
        if [[ $1 -eq 0 || $1 -eq 255 ]]; then
            message "INFO: Successfully $2"
        else
            message "ERROR:  Exit Code $1 for $2"
            exit $1
        fi
    else
        if [ $1 -eq 0 ]; then
            message "INFO: Successfully $2"
        else
            message "ERROR:  Exit Code $1 for $2"
            exit $1
        fi
    fi
}

function create_logs_path {
    log_path=`dirname $yaml_logfile`

    if [ ! -d ${log_path} ]
    then
        if [ "$debug" == "true" ]; then
            message "DEBUG: Creating Logs Path Because Missing..."
            message "DEBUG: Logs Path ($log_path)"
        fi
        mkdir -p ${log_path}
        exit_check $? "Creating Logs Folder"
    fi
}

function set_region {
    if [ "$debug" == "true" ]; then
        message "DEBUG: Setting Region..."
    fi

    if [ -n "$opt_region" ]; then
       region="$opt_region"
    elif [ -n "$yaml_region" ]; then
       region="$yaml_region"
    elif [ -n "$AWS_REGION" ]; then
       region="$AWS_REGION"
    else
       region="us-west-2"
    fi

    if [ "$debug" == "true" ]; then
        message "DEBUG: Region ($region)"
    fi
}

function run_stack_command {
    # Determine if IAM Capabilities are Required
    if [ "$yaml_capabilityiam" == "true" ]; then
        capability_iam=" --capabilities CAPABILITY_IAM"
    elif [ "$yaml_capabilitynamediam" == "true" ]; then
        capability_iam=" --capabilities CAPABILITY_NAMED_IAM"
    else
        capability_iam=""
    fi

    show_header

    if [ "$yaml_uses3template" == "true" ]; then
        aws_command="aws cloudformation $task_type --profile $yaml_profilename --region $region --stack-name $yaml_stackname $capability_iam --template-url "$yaml_templateurl" --parameters file://$yaml_parametersfilepath"
    else
        aws_command="aws cloudformation $task_type --profile $yaml_profilename --region $region --stack-name $yaml_stackname $capability_iam --template-body file://$yaml_templatelocal --parameters file://$yaml_parametersfilepath"
    fi

    if [ "$debug" == "true" ]; then
        message "INFO: AWS Command ($aws_command)"
    fi

    $aws_command
    exit_check $? "Executed ${ACTION} Stack Command"
    echo ''

    monitor_stack_status
}

function delete_stack_command {
    message 'INFO: Deleting Stack'
    aws cloudformation delete-stack --profile ${yaml_profilename} --region ${region}  --stack-name ${yaml_stackname}
    exit_check $? "Executed Delete Stack Command"

    monitor_delete_stack_status
}

function output_create_complete {
    # If Verbose True then Output all the Create Complete Events for Debugging
    if [ "$yaml_verbose" == "true" ]; then
        message "INFO: CREATE COMPLETE EVENTS..."
        echo ''
        create_complete=$(aws cloudformation describe-stack-events --profile ${yaml_profilename} --region ${region}  --stack-name ${yaml_stackname} --output json --query 'StackEvents[?ResourceStatus==`CREATE_COMPLETE`]')
        message "$create_complete"
        echo ''
    fi
}

function output_create_failed {
    # Output all the Create Failed Events for Debugging
    message "INFO: CREATE FAILED EVENTS..."
    echo ''
    create_failed=$(aws cloudformation describe-stack-events --profile ${yaml_profilename} --region ${region}  --stack-name ${yaml_stackname} --output json --query 'StackEvents[?ResourceStatus==`CREATE_FAILED`]')
    message "$create_failed"
    echo ''
}

function monitor_delete_stack_status {
    triggered_delete=true
    wait_time=${yaml_waittime}
    max_waits=${yaml_maxwaits}
    while :
    do
        STATUS=$(aws cloudformation describe-stacks --profile ${yaml_profilename} --region ${region}  --stack-name "$yaml_stackname" --output text --query 'Stacks[*].StackStatus')
        exit_code=$?
        exit_check ${exit_code} "Executing Status Check"
        if [ ${exit_code} -eq 255 ]; then
            message "STATUS: (DOES NOT EXIST)"
        else
            message "STATUS: (${STATUS})"
        fi

        if [[ "$STATUS" == "DELETE_IN_PROGRESS" && ${count} -lt ${max_waits} ]]; then
            message "INFO: Delete not complete!"
            message "INFO: Attempt $count of $max_waits."
            display_runtime
            message "INFO: Polling again in $wait_time seconds..."
            echo ''
            sleep ${wait_time}
            count=$(( count + 1 ))
        elif [ ${exit_code} -eq 255 ]; then
            delete_successful=true
            message "INFO: Stack Deleted ($yaml_stackname)"
            break
        else
            message 'ERROR: The stack delete has failed.'
            break
        fi
    done
}

function get_stack_status {
    STATUS=$(aws cloudformation describe-stacks --profile ${yaml_profilename} --region ${region} --stack-name "$yaml_stackname" --output text --query 'Stacks[*].StackStatus')
    if [ $? -eq 255 ]; then
        message "STATUS: (DOES NOT EXIST) OR (TOKEN EXPIRED)"
    else
        message "STATUS: ($STATUS)"
    fi
}

function monitor_stack_status {
    wait_time=${yaml_waittime}
    max_waits=${yaml_maxwaits}
    while :
    do
        STATUS=$(aws cloudformation describe-stacks --profile ${yaml_profilename} --region ${region} --stack-name "$yaml_stackname" --output text --query 'Stacks[*].StackStatus')
        exit_check $? "Executing Status Check"
        message "STATUS: ($STATUS)"
        elapsed=$(( ($(date +%s) - ${start_time}) / 60 ))

        if [[ "$STATUS" == "${ACTION}_IN_PROGRESS" && ${count} -lt ${max_waits} ]]; then
            message "INFO: $ACTION stack is not complete!"
            message "INFO: Attempt $count of $max_waits."
            display_runtime
            message "INFO: Polling again in $wait_time seconds..."
            echo ''
            sleep ${wait_time}
            count=$(( count + 1 ))
        elif [ "$STATUS" == "${ACTION}_COMPLETE" ]; then
            message "INFO: $ACTION Completed!"
            successful=true
            break
        elif [ "$STATUS" == "${ACTION}_FAILED" ]; then
            message "ERROR:  $ACTION Failed!"
        elif [ "$STATUS" == "UPDATE_COMPLETE_CLEANUP_IN_PROGRESS" ]; then
            message 'INFO: Cleanup in Progress'
            message "INFO: Attempt $count of $max_waits."
            display_runtime
            message "INFO: Polling again in $wait_time seconds..."
            echo ''
            sleep ${wait_time}
            count=$(( count + 1 ))
        elif [ "$STATUS" == "ROLLBACK_IN_PROGRESS" ]; then
            # If Delete Stack on failures when Creating is True then Delete the Stack after grabbing Events
            if [[ "$task_type" == "create-stack" && "$yaml_deletecreatefailures" == "true" ]]; then
                message 'ERROR:  Failed and Rolling Back!'
                echo ''
                output_create_complete
                output_create_failed
                delete_stack_command
                break
            # Else Fetch Events, but do not Delete the Stack
            else
                output_create_complete
                output_create_failed
                message 'ERROR:  Failed and Rolling Back!'
                message "INFO: Rollback not complete!"
                message "INFO: Attempt $count of $max_waits."
                display_runtime
                message "Polling again in $wait_time seconds..."
                echo ''
                sleep ${wait_time}
                count=$(( count + 1 ))
            fi
        elif [ "$STATUS" == "DELETE_IN_PROGRESS" ]; then
            monitor_delete_stack_status
            break
        elif [ "$STATUS" == "ROLLBACK_COMPLETE" ]; then
            message "INFO: Rollback complete!"
            echo ''
            break
        else
            message 'ERROR: The stack has not create or update has failed.'
            echo ''
            break
        fi
    done
}

function display_runtime {
    elapsed_seconds="$(($(date +%s) - ${start_time}))"
    formatted_runtime=$(printf "%02d minutes %02d seconds\n" "$((elapsed_seconds/60%60))" "$((elapsed_seconds%60))")
    message "RUNTIME: ${formatted_runtime}"
}

# Start Line here so if debugging it's before running any functions.
message "** Start CloudFormation Launcher v$script_version **"
if [ "$debug" == "true" ]; then
    message "-- Debugging Enabled --"
fi

# Start Time
start_time=$(date +%s)

# Read Yaml Properties File
eval $(parse_yaml ${config_file_path} "yaml_")

# Create Logs Folder if Logging
if [ ! "$yaml_nolog" == "true" ]; then
    create_logs_path
fi

# Set AWS Region
set_region

# Run Loop
count=1
if [ "$status" == "true" ]; then
    show_header
    get_stack_status
elif [ "$delete" == "true" ]; then
    ACTION=DELETE
    show_header
    delete_stack_command
else
    if [ "$task_type" == "create-stack" ]; then
        ACTION=CREATE
    elif [ "$task_type" == "update-stack" ]; then
        ACTION=UPDATE
    else
        ACTION=CREATE
    fi
    run_stack_command
fi

# Footer
if [ "$status" == "true" ]; then
    echo ''
else
    # Runtime
    end_time=$(date +%s)
    echo ''
    message "ENDTIME: ($(date))"
    display_runtime
    echo ''

    # Results
    if [[ "$delete_successful" == "true" && "$ACTION" == "DELETE" ]]; then
        message "RESULTS: DELETE SUCCESS!"
        echo ''
        exit 0
    elif [ "$successful" == "true" ]; then
        message "RESULTS: $ACTION SUCCESS!"
        echo ''
        exit 0
    else
        message "ERROR: $ACTION FAILED!"
        echo ''
      exit 1
    fi
fi